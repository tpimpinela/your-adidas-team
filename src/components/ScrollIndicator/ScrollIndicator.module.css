.scroll-indicator {
  position: fixed;
  left: 0;
  bottom: 15%;
  margin-left: 0.5rem;
  border-radius: 6px;
  transform: translateX(-100%);
  opacity: 0;
  transition: transform 0.4s, opacity 0.4s;

  &.scroll-indicator--shown {
    transform: unset;
    opacity: 1;
  }

  & button {
    font-size: 1rem;
    padding: 0;
    width: 42px;
    height: 42px;
  }

  & .scroll-indicator__icon {
    display: block;
    transform: rotate(90deg);
  }
}
