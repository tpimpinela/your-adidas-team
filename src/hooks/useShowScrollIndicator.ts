import { useCallback, useEffect, useState } from "react";

const useShowScrollIndicator = (scrollPercentage = 15) => {
  const [showScrollIndicator, setShowScrollIndicator] = useState(false);

  const handleScroll = useCallback(() => {
    const hasScroll =
      document.documentElement.scrollHeight >
      document.documentElement.clientHeight;
    if (!hasScroll) return;
    const maxScroll =
      document.documentElement.scrollHeight -
      document.documentElement.clientHeight;
    setShowScrollIndicator(
      document.documentElement.scrollTop >= maxScroll * (scrollPercentage / 100)
    );
  }, [setShowScrollIndicator, scrollPercentage]);

  useEffect(() => {
    window?.addEventListener("scroll", handleScroll);

    return () => window?.removeEventListener("scroll", handleScroll);
  }, []);

  return showScrollIndicator;
};

export default useShowScrollIndicator;
