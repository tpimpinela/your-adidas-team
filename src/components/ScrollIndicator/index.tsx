import useShowScrollIndicator from "../../hooks/useShowScrollIndicator";
import Button from "../Button";
import styles from "./ScrollIndicator.module.css";

const handleClick = () => {
  window?.scrollTo({ top: 0, behavior: "smooth" });
};

const ScrollIndicator = () => {
  const showScrollIndicator = useShowScrollIndicator();

  return (
    <div
      data-testid="scroll-indicator"
      className={`${styles["scroll-indicator"]} ${
        showScrollIndicator && styles["scroll-indicator--shown"]
      }`}
    >
      <Button onClick={handleClick}>
        <span className={styles["scroll-indicator__icon"]}>&#xab;</span>
      </Button>
    </div>
  );
};

export default ScrollIndicator;
