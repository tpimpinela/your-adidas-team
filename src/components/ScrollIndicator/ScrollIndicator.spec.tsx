import { render, RenderResult, waitFor } from "@testing-library/react";
import { beforeEach, describe, Mock, vi, it, expect } from "vitest";
import ScrollIndicator from ".";
import useShowScrollIndicator from "../../hooks/useShowScrollIndicator";

vi.mock("../../hooks/useShowScrollIndicator", () => ({
  default: vi.fn(),
}));

describe("<ScrollIndicator />", () => {
  let renderResult: RenderResult;
  const useShowScrollIndicatorMock = useShowScrollIndicator as Mock<any, any>;

  beforeEach(() => {
    renderResult = render(<ScrollIndicator />);
  });

  it("should match the snapshot", () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it("should hide indicator", async () => {
    useShowScrollIndicatorMock.mockReturnValue(false);
    renderResult.rerender(<ScrollIndicator />);
    await waitFor(() =>
      expect(
        renderResult
          .getByTestId("scroll-indicator")
          .classList.toString()
          .includes("shown")
      ).toBeFalsy()
    );
  });

  it("should show indicator", async () => {
    useShowScrollIndicatorMock.mockReturnValue(true);
    renderResult.rerender(<ScrollIndicator />);
    await waitFor(() =>
      expect(
        renderResult
          .getByTestId("scroll-indicator")
          .classList.toString()
          .includes("shown")
      ).toBeTruthy()
    );
  });
});
