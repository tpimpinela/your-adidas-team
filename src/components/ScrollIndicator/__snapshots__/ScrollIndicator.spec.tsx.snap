// Vitest Snapshot v1

exports[`<ScrollIndicator /> > should match the snapshot 1`] = `
<div>
  <div
    class="_scroll-indicator_c19ed3 undefined"
    data-testid="scroll-indicator"
  >
    <button
      class="_button_2a1efe"
      data-testid="button-component"
    >
      <span
        class="_scroll-indicator__icon_c19ed3"
      >
        Â«
      </span>
    </button>
  </div>
</div>
`;
